// react-sidebar-styled/dist/index.es.js
import styled from 'styled-components';
import React, { Component } from 'react';
import { PropTypes } from 'prop-types';

var classCallCheck = function (instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
};

var createClass = function () {
  function defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
      var descriptor = props[i];
      descriptor.enumerable = descriptor.enumerable || false;
      descriptor.configurable = true;
      if ("value" in descriptor) descriptor.writable = true;
      Object.defineProperty(target, descriptor.key, descriptor);
    }
  }

  return function (Constructor, protoProps, staticProps) {
    if (protoProps) defineProperties(Constructor.prototype, protoProps);
    if (staticProps) defineProperties(Constructor, staticProps);
    return Constructor;
  };
}();







var _extends = Object.assign || function (target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i];

    for (var key in source) {
      if (Object.prototype.hasOwnProperty.call(source, key)) {
        target[key] = source[key];
      }
    }
  }

  return target;
};



var inherits = function (subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });
  if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
};











var possibleConstructorReturn = function (self, call) {
  if (!self) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return call && (typeof call === "object" || typeof call === "function") ? call : self;
};









var taggedTemplateLiteral = function (strings, raw) {
  return Object.freeze(Object.defineProperties(strings, {
    raw: {
      value: Object.freeze(raw)
    }
  }));
};

var _templateObject = taggedTemplateLiteral(['\n  height: 100%;\n  min-height: 100vh;\n  width: 100%;\n  .sidebar__container {\n    perspective: 1500px;\n    transition: ', ';\n  }\n  .sidebar__menu {\n    position: fixed;\n    right: ', ';\n    transform-style: preserve-3d;\n    transition: inherit;\n    height: 100vh;\n  }\n  .sidebar__page {\n    max-height: ', ';\n    overflow: ', ';\n    transition: ', ';\n  }\n  .sidebar__overlay {\n    max-height: inherit;\n    position: fixed;\n    top: 0;\n    right: 0;\n    bottom: 0;\n    left: 0;\n    background-color: rgb(32, 35, 42);\n    display: ', '\n    opacity: ', ';\n    transition: ', ';\n  }\n  .sidebar__open {\n    position: absolute;\n    top: 0;\n    left: ', ';\n    right: ', ';\n  }\n  .sidebar__close {\n    position: relative;\n    display: flex;\n    justify-content: flex-end;\n    align-items: center;\n    width: 100%;\n    height: 60px;\n    padding-right: 10px;\n    box-sizing: border-box;\n  }\n  .sidebar__close-btn, .sidebar__open-btn {\n    display: block;\n    background: transparent;\n    position: relative;\n    padding: 0;\n    border: 0;\n    &:hover {\n      cursor: pointer;\n    }\n    &:focus {\n      outline: none;\n    }\n  }\n'], ['\n  height: 100%;\n  min-height: 100vh;\n  width: 100%;\n  .sidebar__container {\n    perspective: 1500px;\n    transition: ', ';\n  }\n  .sidebar__menu {\n    position: fixed;\n    right: ', ';\n    transform-style: preserve-3d;\n    transition: inherit;\n    height: 100vh;\n  }\n  .sidebar__page {\n    max-height: ', ';\n    overflow: ', ';\n    transition: ', ';\n  }\n  .sidebar__overlay {\n    max-height: inherit;\n    position: fixed;\n    top: 0;\n    right: 0;\n    bottom: 0;\n    left: 0;\n    background-color: rgb(32, 35, 42);\n    display: ', '\n    opacity: ', ';\n    transition: ', ';\n  }\n  .sidebar__open {\n    position: absolute;\n    top: 0;\n    left: ', ';\n    right: ', ';\n  }\n  .sidebar__close {\n    position: relative;\n    display: flex;\n    justify-content: flex-end;\n    align-items: center;\n    width: 100%;\n    height: 60px;\n    padding-right: 10px;\n    box-sizing: border-box;\n  }\n  .sidebar__close-btn, .sidebar__open-btn {\n    display: block;\n    background: transparent;\n    position: relative;\n    padding: 0;\n    border: 0;\n    &:hover {\n      cursor: pointer;\n    }\n    &:focus {\n      outline: none;\n    }\n  }\n']);

var sidebarBase = styled.div(_templateObject, function (props) {
  return 'transform ' + props.speed + 'ms ' + props.timing;
}, function (props) {
  return props.side === 'right' ? '0' : '';
}, function (props) {
  return props.isOpen || props.active ? '100vh' : '';
}, function (props) {
  return props.isOpen || props.active ? 'hidden' : '';
}, function (props) {
  return 'transform ' + props.speed + 'ms ' + props.timing;
}, function (props) {
  return !props.overlay || !props.isOpen ? 'none' : 'block';
}, function (props) {
  return props.isOpen ? '.75' : '0';
}, function (props) {
  return 'opacity ' + props.speed + 'ms ' + props.timing;
}, function (props) {
  return props.side === 'right' ? '' : '40px';
}, function (props) {
  return props.side === 'right' ? '40px' : '';
});

var _templateObject$1 = taggedTemplateLiteral(['\n  width: 30px;\n  height: 30px;\n  display: block;\n  position: relative;\n  span {\n    position: absolute;\n    top: 5px;\n    display: inline-block;\n    background: white;\n    width: 4px;\n    height: 16px;\n    transform: rotate(45deg);\n    &:last-child {\n      transform: rotate(-45deg);\n    }\n  }\n'], ['\n  width: 30px;\n  height: 30px;\n  display: block;\n  position: relative;\n  span {\n    position: absolute;\n    top: 5px;\n    display: inline-block;\n    background: white;\n    width: 4px;\n    height: 16px;\n    transform: rotate(45deg);\n    &:last-child {\n      transform: rotate(-45deg);\n    }\n  }\n']);

var Styled = styled.span(_templateObject$1);

var CloseIcon = function CloseIcon() {
  return React.createElement(
    Styled,
    { className: 'sidebar__close-icon' },
    React.createElement('span', null),
    React.createElement('span', null)
  );
};

var _templateObject$2 = taggedTemplateLiteral(['\n  display: flex;\n  flex-direction: column;\n  justify-content: space-between;\n  width: 30px;\n  height: 26px;\n  position: relative;\n  span {\n    background-color: white;\n    display: inline-block;\n    width: 30x;\n    height: 4px;\n  }\n'], ['\n  display: flex;\n  flex-direction: column;\n  justify-content: space-between;\n  width: 30px;\n  height: 26px;\n  position: relative;\n  span {\n    background-color: white;\n    display: inline-block;\n    width: 30x;\n    height: 4px;\n  }\n']);

var Styled$1 = styled.span(_templateObject$2);

var OpenIcon = function OpenIcon() {
  return React.createElement(
    Styled$1,
    { className: 'sidebar__open-icon' },
    React.createElement('span', null),
    React.createElement('span', null),
    React.createElement('span', null)
  );
};

var propTypes$1 = {
  side: PropTypes.oneOf(['left', 'right']),
  speed: PropTypes.number,
  timing: PropTypes.string,
  overlay: PropTypes.bool,
  icons: PropTypes.arrayOf(PropTypes.func)
};
var defaultProps = {
  side: 'left',
  speed: 500,
  timing: 'ease',
  overlay: true,
  icons: [function () {
    return React.createElement(CloseIcon, null);
  }, function () {
    return React.createElement(OpenIcon, null);
  }]
};

var sidebarFactory = (function (Styled) {
  var SideBar = function (_Component) {
    inherits(SideBar, _Component);

    function SideBar(props) {
      classCallCheck(this, SideBar);

      var _this = possibleConstructorReturn(this, (SideBar.__proto__ || Object.getPrototypeOf(SideBar)).call(this, props));

      _this.handleSlide = function () {
        clearTimeout(_this.timeout);
        _this.timeout = setTimeout(function () {
          _this.setState(function (prevState, props) {
            return { active: !prevState.isOpen ? false : true };
          });
        }, _this.props.speed);
      };

      _this.toggleMenu = function () {
        _this.setState({ isOpen: !_this.state.isOpen, active: true }, _this.handleSlide());
      };

      _this.state = {
        isOpen: false,
        active: false,
        menuWidth: false
      };
      _this.menuRef = React.createRef();
      return _this;
    }

    createClass(SideBar, [{
      key: 'componentDidMount',
      value: function componentDidMount() {
        this.setState({ menuWidth: this.menuRef.current.clientWidth + 'px' });
      }
    }, {
      key: 'render',
      value: function render() {
        return React.createElement(
          Styled,
          _extends({ className: 'sidebar' }, this.state, this.props),
          React.createElement(
            'div',
            { className: 'sidebar__container' },
            React.createElement(
              'div',
              { className: 'sidebar__menu', ref: this.menuRef },
              React.createElement(
                'div',
                { className: 'sidebar__close' },
                React.createElement(
                  'button',
                  {
                    className: 'sidebar__close-btn',
                    onClick: this.toggleMenu
                  },
                  this.props.icons[0]()
                )
              ),
              this.props.children[0]
            ),
            React.createElement(
              'div',
              { className: 'sidebar__page' },
              this.props.children[1],
              React.createElement(
                'div',
                { className: 'sidebar__open' },
                React.createElement(
                  'button',
                  { className: 'sidebar__open-btn', onClick: this.toggleMenu },
                  this.props.icons[1]()
                )
              ),
              React.createElement('div', { className: 'sidebar__overlay', onClick: this.toggleMenu })
            )
          )
        );
      }
    }]);
    return SideBar;
  }(Component);

  SideBar.propTypes = propTypes$1;
  SideBar.defaultProps = defaultProps;

  return SideBar;
});

var _templateObject$3 = taggedTemplateLiteral(['\n  .sidebar__menu {\n    z-index: -1;\n    transform: ', ';\n  }\n  .sidebar__page {\n    transform: ', ';\n  }\n'], ['\n  .sidebar__menu {\n    z-index: -1;\n    transform: ', ';\n  }\n  .sidebar__page {\n    transform: ', ';\n  }\n']);

var Styled$2 = sidebarBase.extend(_templateObject$3, function (props) {
  return !props.isOpen ? props.side === 'right' ? 'translate3d(-100%, 0, 0)' : 'translate3d(100%, 0, 0)' : 'translate3d(0, 0, 0)';
}, function (props) {
  return !props.isOpen ? 'translate3d(0, 0, 0)' : props.side === 'right' ? 'translate3d(-' + props.menuWidth + ', 0, 0)' : 'translate3d(' + props.menuWidth + ', 0, 0)';
});

var _diverge = sidebarFactory(Styled$2);

var _templateObject$4 = taggedTemplateLiteral(['\n  .sidebar__menu {\n    z-index: -1;\n    transform: ', ';\n  }\n  .sidebar__page {\n    transform: ', ';\n  }\n'], ['\n  .sidebar__menu {\n    z-index: -1;\n    transform: ', ';\n  }\n  .sidebar__page {\n    transform: ', ';\n  }\n']);

var Styled$3 = sidebarBase.extend(_templateObject$4, function (props) {
  return !props.isOpen ? props.side === 'right' ? 'translate3d(0, -100%, 0)' : 'translate3d(0, -100%, 0)' : 'translate3d(0, 0, 0)';
}, function (props) {
  return !props.isOpen ? 'translate3d(0, 0, 0)' : props.side === 'right' ? 'translate3d(-' + props.menuWidth + ', 0, 0)' : 'translate3d(' + props.menuWidth + ', 0, 0)';
});

var _fall = sidebarFactory(Styled$3);

var _templateObject$5 = taggedTemplateLiteral(['\n  .sidebar__container {\n    perspective-origin: ', ';\n  }\n  .sidebar__menu {\n    transform: ', ';\n  }\n  .sidebar__page {\n    transform: ', ';\n  }\n'], ['\n  .sidebar__container {\n    perspective-origin: ', ';\n  }\n  .sidebar__menu {\n    transform: ', ';\n  }\n  .sidebar__page {\n    transform: ', ';\n  }\n']);

var Styled$4 = sidebarBase.extend(_templateObject$5, function (props) {
  return props.side === 'right' ? '100% 50%' : '0% 50%';
}, function (props) {
  return !props.isOpen ? 'translate3d(0, 0, -' + props.menuWidth + ')' : 'translate3d(0, 0, 0)';
}, function (props) {
  return !props.isOpen ? 'translate3d(0, 0, 0)' : props.side === 'right' ? 'translate3d(-' + props.menuWidth + ', 0, 0)' : 'translate3d(' + props.menuWidth + ', 0, 0)';
});

var _grow = sidebarFactory(Styled$4);

var _templateObject$6 = taggedTemplateLiteral(['\n  .sidebar__menu {\n    z-index: 1;\n    transform: ', ';\n  }\n  .sidebar__page {\n    transform: ', ';\n    transform-origin: ', ';\n  }\n'], ['\n  .sidebar__menu {\n    z-index: 1;\n    transform: ', ';\n  }\n  .sidebar__page {\n    transform: ', ';\n    transform-origin: ', ';\n  }\n']);

var Styled$5 = sidebarBase.extend(_templateObject$6, function (props) {
  return !props.isOpen ? props.side === 'right' ? 'translate3d(100%, 0, 0)' : 'translate3d(-100%, 0, 0)' : 'translate3d(0, 0, 0)';
}, function (props) {
  return !props.isOpen ? 'rotateY(0)' : props.side === 'right' ? 'rotateY(10deg)' : 'rotateY(-10deg)';
}, function (props) {
  return props.side === 'right' ? '0% 50%' : '100% 50%';
});

var _press = sidebarFactory(Styled$5);

var _templateObject$7 = taggedTemplateLiteral(['\n  .sidebar__container {\n    transform: ', ';\n  }\n  .sidebar__menu {\n    transform: ', ';\n  }\n'], ['\n  .sidebar__container {\n    transform: ', ';\n  }\n  .sidebar__menu {\n    transform: ', ';\n  }\n']);

var Styled$6 = sidebarBase.extend(_templateObject$7, function (props) {
  return !props.isOpen ? 'translate3d(0, 0, 0)' : props.side === 'right' ? 'translate3d(-' + props.menuWidth + ', 0, 0)' : 'translate3d(' + props.menuWidth + ', 0, 0)';
}, function (props) {
  return props.side === 'right' ? 'translate3d(100%, 0, 0)' : 'translate3d(-100%, 0, 0)';
});

var _push = sidebarFactory(Styled$6);

var _templateObject$8 = taggedTemplateLiteral(['\n  .sidebar__menu {\n    z-index: -1;\n  }\n  .sidebar__page {\n    transform: ', ';\n  }\n'], ['\n  .sidebar__menu {\n    z-index: -1;\n  }\n  .sidebar__page {\n    transform: ', ';\n  }\n']);

var Styled$7 = sidebarBase.extend(_templateObject$8, function (props) {
  return !props.isOpen ? 'translate3d(0, 0, 0)' : props.side === 'right' ? 'translate3d(-' + props.menuWidth + ', 0, 0)' : 'translate3d(' + props.menuWidth + ', 0, 0)';
});

var _reveal = sidebarFactory(Styled$7);

var _templateObject$9 = taggedTemplateLiteral(['\n  .sidebar__menu {\n    z-index: 1;\n    transform: ', ';\n  }\n  .sidebar__page {\n    transform: ', ';\n  }\n'], ['\n  .sidebar__menu {\n    z-index: 1;\n    transform: ', ';\n  }\n  .sidebar__page {\n    transform: ', ';\n  }\n']);

var Styled$8 = sidebarBase.extend(_templateObject$9, function (props) {
  return !props.isOpen ? props.side === 'right' ? 'translate3d(100%, 0, 0)' : 'translate3d(-100%, 0, 0)' : 'translate3d(0, 0, 0)';
}, function (props) {
  return !props.isOpen ? 'translate3d(0, 0, 0)' : 'translate3d(0, 0, -' + props.menuWidth + ')';
});

var _shrink = sidebarFactory(Styled$8);

var _templateObject$10 = taggedTemplateLiteral(['\n  .sidebar__menu {\n    z-index: 1;\n    transform: ', ';\n  }\n'], ['\n  .sidebar__menu {\n    z-index: 1;\n    transform: ', ';\n  }\n']);

var Styled$9 = sidebarBase.extend(_templateObject$10, function (props) {
  return !props.isOpen ? props.side === 'right' ? 'translate3d(100%, 0, 0)' : 'translate3d(-100%, 0, 0)' : 'translate3d(0, 0, 0)';
});

var _slide = sidebarFactory(Styled$9);

var _templateObject$11 = taggedTemplateLiteral(['\n  .sidebar__menu {\n    z-index: -1;\n    transform: ', ';\n  }\n  .sidebar__page {\n    transform: ', ';\n  }\n'], ['\n  .sidebar__menu {\n    z-index: -1;\n    transform: ', ';\n  }\n  .sidebar__page {\n    transform: ', ';\n  }\n']);

var Styled$10 = sidebarBase.extend(_templateObject$11, function (props) {
  return !props.isOpen ? props.side === 'right' ? 'translate3d(50%, 0, 0)' : 'translate3d(-50%, 0, 0)' : 'translate3d(0, 0, 0)';
}, function (props) {
  return !props.isOpen ? 'translate3d(0, 0, 0)' : props.side === 'right' ? 'translate3d(-' + props.menuWidth + ', 0, 0)' : 'translate3d(' + props.menuWidth + ', 0, 0)';
});

var _uncover = sidebarFactory(Styled$10);

export { _diverge as diverge, _fall as fall, _grow as grow, _press as press, _push as push, _reveal as reveal, _shrink as shrink, _slide as slide, _uncover as uncover };
